File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating.java
import java.util.ArrayList;

class Solution {
    char[] sArray;

    public int minFlips(String s) {

        sArray = s.toCharArray();

        ArrayList<Integer> start0 = new ArrayList<>();
        ArrayList<Integer> start1 = new ArrayList<>();

        for (int i = 0; i < sArray.length; i++) {
            if ((i & 1) == 1) {
                if (sArray[i] == '0') {
                    start0.add(i);
                } else {
                    start1.add(i);
                }
            } else {
                if (sArray[i] == '0') {
                    start1.add(i);
                } else {
                    start0.add(i);
                }
            }
        }

        return Math.min(brutalForce(start0), brutalForce(start1));

    }

    private int brutalForce(ArrayList<Integer> placeToSwitch) {
        if (placeToSwitch.size() == 0)
            return 0;
        // printArrayList(placeToSwitch);
        int ans = placeToSwitch.size();
        int sum = ans;

        int last = sArray[sArray.length - 1];
        if (placeToSwitch.get(placeToSwitch.size() - 1) == sArray.length - 1) {
            last = last == '0' ? '1' : '0';
        }

        int j = 0;
        for (int i = 0; i < sArray.length && j < placeToSwitch.size(); i++) {
            if (sArray[i] == last) {
                sum++;
            }

            last = last == '0' ? '1' : '0';

            if (i == placeToSwitch.get(j)) {
                sum--;
                ans = Math.min(ans, sum);
                j++;
            }
        }
        return ans;
    }

    private void printArrayList(ArrayList<Integer> object) {
        object.forEach(value -> System.out.print(value + "->"));
        System.out.println("");
    }
}

File: python/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.py
class Solution:
    def minFlips(self, s: str) -> int:
        N = len(s)
        l, counter_zero, counter_one, start_zero, start_one, ans, sList = 0, 0, 0, [0]*2*N, [0]*2*N, N, [int(i) for i in s]*2
        start_zero[0], start_one[0] = 0, 1
        
        for i in range(1, 2*N):
            start_zero[i] = 1 if start_zero[i-1]==0 else 0
            start_one[i] = 1 if start_one[i-1]==0 else 0

        for r in range(2*N):
            counter_zero +=1 if start_zero[r]!= sList[r] else 0
            counter_one +=1 if start_one[r]!= sList[r] else 0
            
            if r==N-1:
                ans = min(ans, counter_zero, counter_one)

            if r >= N:
                counter_zero -=1 if start_zero[l]!= sList[l] else 0
                counter_one -=1 if start_one[l]!= sList[l] else 0
                ans = min(ans, counter_zero, counter_one)
                l+=1
                
                
        return ans
            
            
        
        
        
        